pkg/publicshare/manager/memory/memory.go

```diff
@@ -34,50 +37,145 @@ func init() {
 	registry.Register("memory", New)
 }
 
-// New returns a new memory manager
-func New(c map[string]interface{}) (publicshare.Manager, error) {
-	return &manager{}, nil
+// New returns a new memory manager.
+func New() (publicshare.Manager, error) {
+	return &manager{
+		shares: sync.Map{},
+	}, nil
 }
 
-type manager struct{}
+type manager struct {
+	shares sync.Map
+}
 
-// TODO(refs) implement application logic.
-func (m *manager) CreatePublicShare(ctx context.Context, u *authv0alphapb.User, md *storageproviderv0alphapb.ResourceInfo, g *publicshareproviderv0alphapb.Grant) (*publicshareproviderv0alphapb.PublicShare, error) {
-	return &publicshareproviderv0alphapb.PublicShare{}, nil
+// CreatePublicShare safely adds a new entry to manager.shares
+func (m *manager) CreatePublicShare(ctx context.Context, u *authv0alphapb.User, rInfo *storageproviderv0alphapb.ResourceInfo, g *publicshareproviderv0alphapb.Grant) (*publicshareproviderv0alphapb.PublicShare, error) {
+	// what to set id?
+	//  TODO does OpaqueId expecting a specific format?
+	// where could this initialization go wrong and early return?
+	id := &publicshareproviderv0alphapb.PublicShareId{
+		OpaqueId: randString(12),
+	}
+	tkn := randString(12)
+	now := uint64(time.Now().Unix())
+
+	newShare := publicshareproviderv0alphapb.PublicShare{
+		Id:          id,
+		Owner:       u.Id,
+		Creator:     u.Id,
+		ResourceId:  rInfo.Id,
+		Token:       tkn,
+		Permissions: g.Permissions,
+		Ctime: &typespb.Timestamp{
+			Seconds: now,
+			Nanos:   uint32(now % 1000000000),
+		},
+		Mtime: &typespb.Timestamp{
+			Seconds: now,
+			Nanos:   uint32(now % 1000000000),
+		},
+		PasswordProtected: false,
+		Expiration:        g.Expiration,
+		DisplayName:       tkn,
+	}
+
+	m.shares.Store(newShare.Token, &newShare)
+	return &newShare, nil
 }
 
+// UpdatePublicShare updates the expiration date, permissions and Mtime
 func (m *manager) UpdatePublicShare(ctx context.Context, u *authv0alphapb.User, ref *publicshareproviderv0alphapb.PublicShareReference, g *publicshareproviderv0alphapb.Grant) (*publicshareproviderv0alphapb.PublicShare, error) {
+	share, err := m.GetPublicShare(ctx, u, ref)
+	if err != nil {
+		return nil, errors.New("ref does not exist")
+	}
+
+	token := share.GetToken()
+
+	// thread unsafe. 2 goroutines can access to the same resource?
+	share.Permissions = g.Permissions
+	share.Expiration = g.Expiration
+	share.Mtime = &typespb.Timestamp{
+		Seconds: uint64(time.Now().Unix()),
+		Nanos:   uint32(time.Now().Unix() % 1000000000),
+	}
+
+	m.shares.Store(token, share)
```

Comment:
I wonder if we should update the share Creator ... it would then always reflect who last changed the share properties. That allows answering the question whe is responsible for the share. @labkode opinion?

Reply:
if we do this, then creator will lose its meaning as "the one who originally created the share" and if something breaks / leaks in the future the information of the original creator is lost. 

Wouldn't it be better to add a new attribute? something along the lines of `UpdatedBy`

Reply:
yeah, I was worrying about audits as well. But it isnt really important who initially created the share ... just who last updated it. Because that person is responsible for the current share permissions. The one to blame, so to say. Ultimately, the owner of the file is responsible to keep track of who has access to his files, which is why there is the reshare (or manage acl) permission. Furthermore, an audit log will keep track of all changes. That also includes who initially shared the file. ... And thinking about it ... that will always be the owner. Even if the resource is owned by a group, then all members are responsible for managing the sharing permissions. I think updating the share Creator is the right thing to do. I agree that the name should be changed. Maybe `Editor`?

Reply:
If so, this is a CS3API chinge request. @labkode should we open an issue in the cs3API? Or should @refs just submit a PR?